aten/src/ATen/native/cpu/MaxUnpoolKernel.cpp

```diff
@@ -9,6 +9,25 @@ namespace at { namespace native {
 
 namespace {
 
+template<typename T>
+class atomic_optional {
+public:
+  atomic_optional() = default;
+  operator bool() const {
+    return valid_.load();
+  }
+  T value() const {
+    return val_.load();
+  }
+  void operator=(const T& val) {
+    valid_ = true;
+    val_ = val;
+  }
+private:
+  std::atomic<T> val_{};
+  std::atomic<T> valid_{false};
```

Comment:
looks like only `valid_` should be atomic, `val_` (index) is only needed for error message and any thread could write it?
Come to think of it, even `valid_` doesn't need to be atomic, if multiple threads set it to true that's fine (and no thread can set it to false)

Reply:
Good point, so perhaps regular c10::optional + thread_fence should be sufficient. I wonder if I should split this PR into two: one that uses c10::optional and keeps `omp critical` pragma?

Reply:
I think it can be done in one PR, but up to you